File: frontend/src/services/robust.ts
import { http } from './http'

export type RobustJobStatus = 'queued' | 'running' | 'success' | 'failed'

export type RobustJob = {
  id: string
  status: RobustJobStatus
  modelId: string
  promptId: string
  attackConfigJson: string
  metricsJson: string
  error: string
  startedAt?: string | null
  finishedAt?: string | null
  createdAt: string
  updatedAt: string
}

export type RobustMetrics = {
  targetVulnType?: string
  attackMethod?: string
  attackPipeline?: string[]
  victimResults?: Array<{
    victimModel: string
    displayName: string
    supported: boolean
    skippedReason?: string
    cacheHit?: boolean
    targetVulnType?: string
    attackableContracts: number
    totalAdversarial: number
    attackSuccesses: number
    attackSuccessRate: number
    origAccuracy: number
    advAccuracy: number
    accuracyDropRate: number
    avgConfidenceDrop: number
    avgQueries: number
    avgPerturbationRate: number
    avgVisiblePerturbationRate: number
    avgCodeBLEU: number
    queryBudgetHits: number
    visibilityWarning?: string
    perStrategy?: Array<{
      strategy: string
      totalVariants: number
      attackSuccesses: number
      attackSuccessRate: number
      avgConfidenceDrop: number
      avgQueries: number
      avgPerturbationRate: number
      avgVisiblePerturbationRate: number
      avgCodeBLEU: number
      queryBudgetHits: number
    }>
    perContract?: Array<{
      baseContractId: string
      contractName: string
      origLabel: string
      origConfidence: number
      origVulnScore: number
      attackable: boolean
      skippedReason?: string
      advTotal: number
      flipped: number
      avgAdvConfidence: number
      avgConfDrop: number
      avgQueries: number
      avgPerturbationRate: number
      avgVisiblePerturbationRate: number
      avgCodeBLEU: number
      queryBudgetHits: number
      bestAttackStrategy?: string
      bestAttackSample?: {
        variantIndex: number
        queries: number
        perturbationTokens: number
        originalTokens: number
        perturbationRate: number
        visiblePerturbationRate: number
        codebleu: number
        queryBudgetHit: boolean
        confidenceDrop: number
        vulnScoreDrop: number
        attackSucceeded: boolean
      }
      byStrategy?: Record<
        string,
        {
          total: number
          attackSuccesses: number
          attackSuccessRate: number
          avgConfidenceDrop: number
          avgQueries: number
          avgPerturbationRate: number
          avgVisiblePerturbationRate: number
          avgCodeBLEU: number
          queryBudgetHits: number
        }
      >
    }>
  }>
  attackableContracts?: number
  totalAdversarial?: number
  attackSuccesses?: number
  attackSuccessRate?: number
  origAccuracy?: number
  advAccuracy?: number
  accuracyDropRate?: number
  avgConfidenceDrop?: number
  avgQueries?: number
  avgPerturbationRate?: number
  avgVisiblePerturbationRate?: number
  avgCodeBLEU?: number
  queryBudgetHits?: number
  inputWindowMaxLength?: number
  visibilityWarning?: string
  perStrategy?: Array<{
    strategy: string
    totalVariants: number
    attackSuccesses: number
    attackSuccessRate: number
    avgConfidenceDrop: number
    avgQueries: number
    avgPerturbationRate: number
    avgVisiblePerturbationRate: number
    avgCodeBLEU: number
    queryBudgetHits: number
  }>
  perContract?: Array<{
    baseContractId: string
    contractName: string
    origLabel: string
    origConfidence: number
    origVulnScore: number
    attackable: boolean
    skippedReason?: string
    advTotal: number
    flipped: number
    avgAdvConfidence: number
    avgConfDrop: number
    avgQueries: number
    avgPerturbationRate: number
    avgVisiblePerturbationRate: number
    avgCodeBLEU: number
    queryBudgetHits: number
    bestAttackStrategy?: string
    bestAttackSample?: {
      variantIndex: number
      queries: number
      perturbationTokens: number
      originalTokens: number
      perturbationRate: number
      visiblePerturbationRate: number
      codebleu: number
      queryBudgetHit: boolean
      confidenceDrop: number
      vulnScoreDrop: number
      attackSucceeded: boolean
    }
    byStrategy?: Record<
      string,
      {
        total: number
        attackSuccesses: number
        attackSuccessRate: number
        avgConfidenceDrop: number
        avgQueries: number
        avgPerturbationRate: number
        avgVisiblePerturbationRate: number
        avgCodeBLEU: number
        queryBudgetHits: number
      }
    >
  }>
  [key: string]: unknown
}

function asNumber(value: unknown, fallback = 0) {
  return typeof value === 'number' && Number.isFinite(value) ? value : fallback
}

function asString(value: unknown, fallback = '') {
  return typeof value === 'string' ? value : fallback
}

function asBool(value: unknown, fallback = false) {
  return typeof value === 'boolean' ? value : fallback
}

function notNull<T>(value: T | null): value is T {
  return value !== null
}

function normalizeBestAttackSample(sample: any) {
  if (!sample || typeof sample !== 'object') return undefined
  return {
    variantIndex: asNumber(sample.variantIndex),
    queries: asNumber(sample.queries),
    perturbationTokens: asNumber(sample.perturbationTokens),
    originalTokens: asNumber(sample.originalTokens),
    perturbationRate: asNumber(sample.perturbationRate),
    visiblePerturbationRate: asNumber(sample.visiblePerturbationRate),
    codebleu: asNumber(sample.codebleu),
    queryBudgetHit: asBool(sample.queryBudgetHit),
    confidenceDrop: asNumber(sample.confidenceDrop),
    vulnScoreDrop: asNumber(sample.vulnScoreDrop),
    attackSucceeded: asBool(sample.attackSucceeded),
  }
}

function normalizePerStrategy(strategy: any) {
  if (!strategy || typeof strategy !== 'object') return null
  return {
    strategy: asString(strategy.strategy),
    totalVariants: asNumber(strategy.totalVariants ?? strategy.total),
    attackSuccesses: asNumber(strategy.attackSuccesses),
    attackSuccessRate: asNumber(strategy.attackSuccessRate),
    avgConfidenceDrop: asNumber(strategy.avgConfidenceDrop ?? strategy.avgConfDrop),
    avgQueries: asNumber(strategy.avgQueries),
    avgPerturbationRate: asNumber(strategy.avgPerturbationRate),
    avgVisiblePerturbationRate: asNumber(strategy.avgVisiblePerturbationRate),
    avgCodeBLEU: asNumber(strategy.avgCodeBLEU),
    queryBudgetHits: asNumber(strategy.queryBudgetHits),
  }
}

function normalizeByStrategy(byStrategy: unknown) {
  if (!byStrategy || typeof byStrategy !== 'object') return undefined
  return Object.fromEntries(
    Object.entries(byStrategy as Record<string, any>).map(([key, value]) => [
      key,
      {
        total: asNumber(value?.total ?? value?.totalVariants),
        attackSuccesses: asNumber(value?.attackSuccesses),
        attackSuccessRate: asNumber(value?.attackSuccessRate),
        avgConfidenceDrop: asNumber(value?.avgConfidenceDrop ?? value?.avgConfDrop),
        avgQueries: asNumber(value?.avgQueries),
        avgPerturbationRate: asNumber(value?.avgPerturbationRate),
        avgVisiblePerturbationRate: asNumber(value?.avgVisiblePerturbationRate),
        avgCodeBLEU: asNumber(value?.avgCodeBLEU),
        queryBudgetHits: asNumber(value?.queryBudgetHits),
      },
    ]),
  )
}

function normalizePerContract(item: any) {
  if (!item || typeof item !== 'object') return null
  return {
    baseContractId: asString(item.baseContractId),
    contractName: asString(item.contractName),
    origLabel: asString(item.origLabel),
    origConfidence: asNumber(item.origConfidence),
    origVulnScore: asNumber(item.origVulnScore),
    attackable: asBool(item.attackable),
    skippedReason: typeof item.skippedReason === 'string' ? item.skippedReason : undefined,
    advTotal: asNumber(item.advTotal),
    flipped: asNumber(item.flipped),
    avgAdvConfidence: asNumber(item.avgAdvConfidence),
    avgConfDrop: asNumber(item.avgConfDrop ?? item.avgConfidenceDrop),
    avgQueries: asNumber(item.avgQueries),
    avgPerturbationRate: asNumber(item.avgPerturbationRate),
    avgVisiblePerturbationRate: asNumber(item.avgVisiblePerturbationRate),
    avgCodeBLEU: asNumber(item.avgCodeBLEU),
    queryBudgetHits: asNumber(item.queryBudgetHits),
    bestAttackStrategy: item.bestAttackStrategy ? asString(item.bestAttackStrategy) : undefined,
    bestAttackSample: normalizeBestAttackSample(item.bestAttackSample),
    byStrategy: normalizeByStrategy(item.byStrategy),
  }
}

function normalizeVictimResult(item: any) {
  if (!item || typeof item !== 'object') return null
  const perStrategy = Array.isArray(item.perStrategy)
    ? item.perStrategy.map(normalizePerStrategy).filter(notNull)
    : []
  const perContract = Array.isArray(item.perContract)
    ? item.perContract.map(normalizePerContract).filter(notNull)
    : []
  return {
    victimModel: asString(item.victimModel),
    displayName: asString(item.displayName || item.victimModel),
    supported: item.supported !== false,
    skippedReason: item.skippedReason ? asString(item.skippedReason) : undefined,
    cacheHit: typeof item.cacheHit === 'boolean' ? item.cacheHit : undefined,
    targetVulnType: item.targetVulnType ? asString(item.targetVulnType) : undefined,
    attackableContracts: asNumber(item.attackableContracts),
    totalAdversarial: asNumber(item.totalAdversarial),
    attackSuccesses: asNumber(item.attackSuccesses),
    attackSuccessRate: asNumber(item.attackSuccessRate),
    origAccuracy: asNumber(item.origAccuracy),
    advAccuracy: asNumber(item.advAccuracy),
    accuracyDropRate: asNumber(item.accuracyDropRate),
    avgConfidenceDrop: asNumber(item.avgConfidenceDrop),
    avgQueries: asNumber(item.avgQueries),
    avgPerturbationRate: asNumber(item.avgPerturbationRate),
    avgVisiblePerturbationRate: asNumber(item.avgVisiblePerturbationRate),
    avgCodeBLEU: asNumber(item.avgCodeBLEU),
    queryBudgetHits: asNumber(item.queryBudgetHits),
    visibilityWarning: item.visibilityWarning ? asString(item.visibilityWarning) : undefined,
    perStrategy,
    perContract,
  }
}

export function normalizeRobustMetrics(input?: RobustMetrics | Record<string, unknown> | null): RobustMetrics | null {
  if (!input || typeof input !== 'object') return null
  const raw = input as Record<string, unknown>

  const perStrategy = Array.isArray(raw.perStrategy)
    ? raw.perStrategy.map(normalizePerStrategy).filter(notNull)
    : []
  const perContract = Array.isArray(raw.perContract)
    ? raw.perContract.map(normalizePerContract).filter(notNull)
    : []
  const victimResults = Array.isArray(raw.victimResults)
    ? raw.victimResults.map(normalizeVictimResult).filter(notNull)
    : []

  const normalized: RobustMetrics = {
    ...raw,
    targetVulnType: asString(raw.targetVulnType),
    attackMethod: asString(raw.attackMethod),
    attackPipeline: Array.isArray(raw.attackPipeline)
      ? raw.attackPipeline.filter((x): x is string => typeof x === 'string')
      : [],
    attackableContracts: asNumber(raw.attackableContracts),
    totalAdversarial: asNumber(raw.totalAdversarial),
    attackSuccesses: asNumber(raw.attackSuccesses),
    attackSuccessRate: asNumber(raw.attackSuccessRate),
    origAccuracy: asNumber(raw.origAccuracy),
    advAccuracy: asNumber(raw.advAccuracy),
    accuracyDropRate: asNumber(raw.accuracyDropRate),
    avgConfidenceDrop: asNumber(raw.avgConfidenceDrop),
    avgQueries: asNumber(raw.avgQueries),
    avgPerturbationRate: asNumber(raw.avgPerturbationRate),
    avgVisiblePerturbationRate: asNumber(raw.avgVisiblePerturbationRate),
    avgCodeBLEU: asNumber(raw.avgCodeBLEU),
    queryBudgetHits: asNumber(raw.queryBudgetHits),
    inputWindowMaxLength: asNumber(raw.inputWindowMaxLength),
    visibilityWarning: raw.visibilityWarning ? asString(raw.visibilityWarning) : undefined,
    victimResults,
    perStrategy,
    perContract,
  }

  if (victimResults.length === 0) {
    normalized.victimResults = [
      {
        victimModel: 'codebert',
        displayName: 'CodeBERT',
        supported: true,
        targetVulnType: normalized.targetVulnType,
        attackableContracts: normalized.attackableContracts ?? 0,
        totalAdversarial: normalized.totalAdversarial ?? 0,
        attackSuccesses: normalized.attackSuccesses ?? 0,
        attackSuccessRate: normalized.attackSuccessRate ?? 0,
        origAccuracy: normalized.origAccuracy ?? 0,
        advAccuracy: normalized.advAccuracy ?? 0,
        accuracyDropRate: normalized.accuracyDropRate ?? 0,
        avgConfidenceDrop: normalized.avgConfidenceDrop ?? 0,
        avgQueries: normalized.avgQueries ?? 0,
        avgPerturbationRate: normalized.avgPerturbationRate ?? 0,
        avgVisiblePerturbationRate: normalized.avgVisiblePerturbationRate ?? 0,
        avgCodeBLEU: normalized.avgCodeBLEU ?? 0,
        queryBudgetHits: normalized.queryBudgetHits ?? 0,
        visibilityWarning: normalized.visibilityWarning,
        perStrategy,
        perContract,
      },
    ]
  }

  if (!normalized.attackMethod) {
    normalized.attackMethod = perStrategy[0]?.strategy ?? ''
  }
  return normalized
}

export async function createRobustJob(body: {
  modelId: string
  promptId: string
  contractIds: string[]
  victimModels: string[]
  strategies: string[]
  variantsPerSource: number
}) {
  const { data } = await http.post<RobustJob>('/api/robust/evaluate', body)
  return data
}

export async function getRobustJob(id: string) {
  const { data } = await http.get<{
    job: RobustJob
    metrics?: RobustMetrics
  }>(`/api/robust/jobs/${id}`)
  return {
    ...data,
    metrics: normalizeRobustMetrics(data.metrics),
  }
}

export async function listRobustJobs() {
  const { data } = await http.get<RobustJob[]>('/api/robust/jobs')
  return data
}

File: backend/internal/service/robust.go
package service

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"sc-vuln-detector/backend/internal/model"
)

const (
	RobustStrategyDIPAttack = "dip-attack"
)

type RobustService struct {
	DB       *gorm.DB
	Detector *Detector
}

func NewRobustService(db *gorm.DB, detector *Detector) *RobustService {
	return &RobustService{DB: db, Detector: detector}
}

type RobustEvaluateRequest struct {
	ModelID        string   `json:"modelId"`
	PromptID       string   `json:"promptId"`
	ContractIDs    []string `json:"contractIds"`
	VictimModels   []string `json:"victimModels"`
	Strategies     []string `json:"strategies"`
	VariantsPerSrc int      `json:"variantsPerSource"`
}

type robustConfig struct {
	ContractIDs       []string `json:"contractIds"`
	VictimModels      []string `json:"victimModels"`
	Strategies        []string `json:"strategies"`
	VariantsPerSource int      `json:"variantsPerSource"`
}

type modelArtifactMetadata struct {
	TargetVulnType string         `json:"target_vuln_type"`
	LabelMap       map[string]int `json:"label_map"`
	PromptText     *string        `json:"prompt_text"`
	MaxLength      int            `json:"max_length"`
	RawClassCounts map[string]int `json:"raw_class_counts"`
	TrainCounts    map[string]int `json:"train_counts"`
	ValCounts      map[string]int `json:"val_counts"`
	Extra          map[string]any `json:"-"`
}

type dipInferenceSummary struct {
	Label      string  `json:"label"`
	Confidence float64 `json:"confidence"`
	VulnScore  float64 `json:"vulnScore"`
	ElapsedMS  int     `json:"elapsedMs"`
}

type dipAttackVariant struct {
	VariantIndex              int                 `json:"variantIndex"`
	Success                   bool                `json:"success"`
	Queries                   int                 `json:"queries"`
	QueryBudgetHit            bool                `json:"queryBudgetHit"`
	PerturbationTokens        int                 `json:"perturbationTokens"`
	OriginalTokens            int                 `json:"originalTokens"`
	PerturbationRate          float64             `json:"perturbationRate"`
	VisiblePerturbationTokens int                 `json:"visiblePerturbationTokens"`
	VisibleWindowTokens       int                 `json:"visibleWindowTokens"`
	VisiblePerturbationRate   float64             `json:"visiblePerturbationRate"`
	CodeBLEU                  float64             `json:"codebleu"`
	AdvCode                   string              `json:"advCode"`
	Baseline                  dipInferenceSummary `json:"baseline"`
	Adversarial               dipInferenceSummary `json:"adversarial"`
	ConfidenceDrop            float64             `json:"confidenceDrop"`
	VulnScoreDrop             float64             `json:"vulnScoreDrop"`
	VisibleLineEnd            int                 `json:"visibleLineEnd"`
}

type dipAttackResult struct {
	TargetVulnType        string              `json:"targetVulnType"`
	Baseline              dipInferenceSummary `json:"baseline"`
	Variants              []dipAttackVariant  `json:"variants"`
	Attackable            bool                `json:"attackable"`
	MaxLength             int                 `json:"maxLength"`
	VisibleLineEnd        int                 `json:"visibleLineEnd"`
	OriginalVisibleTokens int                 `json:"originalVisibleTokens"`
	OriginalTotalTokens   int                 `json:"originalTotalTokens"`
}

type robustAttackSample struct {
	VariantIndex            int     `json:"variantIndex"`
	Queries                 int     `json:"queries"`
	PerturbationTokens      int     `json:"perturbationTokens"`
	OriginalTokens          int     `json:"originalTokens"`
	PerturbationRate        float64 `json:"perturbationRate"`
	ConfidenceDrop          float64 `json:"confidenceDrop"`
	VulnScoreDrop           float64 `json:"vulnScoreDrop"`
	VisiblePerturbationRate float64 `json:"visiblePerturbationRate"`
	CodeBLEU                float64 `json:"codebleu"`
	QueryBudgetHit          bool    `json:"queryBudgetHit"`
	AttackSucceeded         bool    `json:"attackSucceeded"`
}

type robustPerContract struct {
	BaseContractID             string              `json:"baseContractId"`
	ContractName               string              `json:"contractName"`
	OrigLabel                  model.Label         `json:"origLabel"`
	OrigConfidence             float64             `json:"origConfidence"`
	OrigVulnScore              float64             `json:"origVulnScore"`
	Attackable                 bool                `json:"attackable"`
	SkippedReason              string              `json:"skippedReason"`
	AdvTotal                   int                 `json:"advTotal"`
	Flipped                    int                 `json:"flipped"`
	AvgAdvConfidence           float64             `json:"avgAdvConfidence"`
	AvgConfDrop                float64             `json:"avgConfDrop"`
	AvgQueries                 float64             `json:"avgQueries"`
	AvgPerturbationRate        float64             `json:"avgPerturbationRate"`
	AvgVisiblePerturbationRate float64             `json:"avgVisiblePerturbationRate"`
	AvgCodeBLEU                float64             `json:"avgCodeBLEU"`
	QueryBudgetHits            int                 `json:"queryBudgetHits"`
	BestAttackStrategy         string              `json:"bestAttackStrategy"`
	BestAttackSample           *robustAttackSample `json:"bestAttackSample,omitempty"`
	ByStrategy                 map[string]any      `json:"byStrategy"`
}

type robustStrategyAgg struct {
	TotalVariants              int
	AttackSuccesses            int
	ConfidenceDropSum          float64
	QuerySumAll                float64
	QuerySumSuccess            float64
	SuccessQueryCount          int
	PerturbationRateSumAll     float64
	PerturbationRateSumSuccess float64
	SuccessPerturbationCount   int
	VisiblePerturbationRateSum float64
	CodeBLEUSum                float64
	QueryBudgetHits            int
}

func (s *RobustService) CreateJob(ctx context.Context, req RobustEvaluateRequest) (*model.RobustJob, error) {
	if s.DB == nil || s.Detector == nil {
		return nil, gorm.ErrInvalidDB
	}
	if strings.TrimSpace(req.ModelID) == "" || strings.TrimSpace(req.PromptID) == "" {
		return nil, fmt.Errorf("modelId 和 promptId 必填")
	}
	if len(req.ContractIDs) == 0 {
		return nil, fmt.Errorf("contractIds 不能为空")
	}
	if len(req.Strategies) == 0 {
		req.Strategies = []string{RobustStrategyDIPAttack}
	}
	if len(req.VictimModels) == 0 {
		req.VictimModels = []string{"codebert", "AME", "GPSCVul", "ConvMHSA", "Clear"}
	}
	if req.VariantsPerSrc <= 0 {
		req.VariantsPerSrc = 1
	}
	if req.VariantsPerSrc > 5 {
		req.VariantsPerSrc = 5
	}

	confBytes, _ := json.Marshal(robustConfig{
		ContractIDs:       req.ContractIDs,
		VictimModels:      req.VictimModels,
		Strategies:        req.Strategies,
		VariantsPerSource: req.VariantsPerSrc,
	})

	job := &model.RobustJob{
		ID:               uuid.NewString(),
		Status:           model.RobustJobStatusQueued,
		ModelID:          req.ModelID,
		PromptID:         req.PromptID,
		AttackConfigJSON: string(confBytes),
	}
	if err := s.DB.WithContext(ctx).Create(job).Error; err != nil {
		return nil, err
	}

	go s.run(job.ID)
	return job, nil
}

func (s *RobustService) run(jobID string) {
	ctx := context.Background()

	var job model.RobustJob
	if err := s.DB.WithContext(ctx).First(&job, "id = ?", jobID).Error; err != nil {
		return
	}

	start := time.Now()
	if err := s.DB.WithContext(ctx).Model(&model.RobustJob{}).
		Where("id = ?", jobID).
		Updates(map[string]any{
			"status":     model.RobustJobStatusRunning,
			"started_at": start,
			"error":      "",
		}).Error; err != nil {
		return
	}

	var trainedModel model.TrainedModel
	if err := s.DB.WithContext(ctx).First(&trainedModel, "id = ?", job.ModelID).Error; err != nil {
		s.fail(jobID, fmt.Errorf("模型不存在"))
		return
	}

	var prompt model.Prompt
	if err := s.DB.WithContext(ctx).First(&prompt, "id = ?", job.PromptID).Error; err != nil {
		s.fail(jobID, fmt.Errorf("提示模板不存在"))
		return
	}

	meta, err := s.loadModelMetadata(trainedModel.Artifact)
	if err != nil {
		s.fail(jobID, err)
		return
	}

	var cfg robustConfig
	if err := json.Unmarshal([]byte(job.AttackConfigJSON), &cfg); err != nil {
		s.fail(jobID, fmt.Errorf("解析任务配置失败: %w", err))
		return
	}
	if len(cfg.Strategies) == 0 {
		cfg.Strategies = []string{RobustStrategyDIPAttack}
	}
	if len(cfg.VictimModels) == 0 {
		cfg.VictimModels = []string{"codebert", "AME", "GPSCVul", "ConvMHSA", "Clear"}
	}

	var contracts []model.Contract
	if err := s.DB.WithContext(ctx).Find(&contracts, "id IN ?", cfg.ContractIDs).Error; err != nil {
		s.fail(jobID, fmt.Errorf("加载合约失败: %w", err))
		return
	}
	if len(contracts) == 0 {
		s.fail(jobID, fmt.Errorf("未找到任何合约"))
		return
	}

	baseIDs := make([]string, 0, len(contracts))
	for _, contract := range contracts {
		baseIDs = append(baseIDs, contract.ID)
	}
	_ = s.DB.WithContext(ctx).
		Where("base_contract_id IN ?", baseIDs).
		Delete(&model.AdversarialSample{}).Error
	metrics, err := s.runMultiVictimRobustness(trainedModel.Artifact, prompt.TemplateText, meta.TargetVulnType, cfg.VictimModels, cfg.VariantsPerSource, contracts)
	if err != nil {
		s.fail(jobID, err)
		return
	}

	finish := time.Now()
	_ = s.DB.WithContext(ctx).Model(&model.RobustJob{}).Where("id = ?", jobID).
		Updates(map[string]any{
			"status":       model.RobustJobStatusSuccess,
			"metrics_json": mustMarshal(metrics),
			"finished_at":  finish,
		}).Error
}

func (s *RobustService) runDIPAttack(artifactPath, source, promptText string, variantsPerSource int) (*dipAttackResult, error) {
	modelDir, err := resolveArtifactPath(artifactPath)
	if err != nil {
		return nil, err
	}

	cmd := exec.Command(
		pythonExecutable(),
		filepath.ToSlash(filepath.Join("..", "python_scripts", "run_dip_attack.py")),
		"--model_dir", modelDir,
		"--variants", fmt.Sprintf("%d", variantsPerSource),
	)
	if strings.TrimSpace(promptText) != "" {
		cmd.Args = append(cmd.Args, "--prompt_text", promptText)
	}
	cmd.Stdin = strings.NewReader(source)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("DIP 攻击执行失败: %w; stderr=%s", err, strings.TrimSpace(stderr.String()))
	}

	var result dipAttackResult
	if err := json.Unmarshal(output, &result); err != nil {
		return nil, fmt.Errorf("解析 DIP 攻击结果失败: %w; stdout=%s; stderr=%s", err, strings.TrimSpace(string(output)), strings.TrimSpace(stderr.String()))
	}
	return &result, nil
}

func (s *RobustService) runMultiVictimRobustness(
	artifactPath, promptText, targetVulnType string,
	victimModels []string,
	variantsPerSource int,
	contracts []model.Contract,
) (map[string]any, error) {
	modelDir, err := resolveArtifactPath(artifactPath)
	if err != nil {
		return nil, err
	}

	payloadContracts := make([]map[string]string, 0, len(contracts))
	for _, contract := range contracts {
		payloadContracts = append(payloadContracts, map[string]string{
			"id":              contract.ID,
			"name":            contract.Name,
			"processedSource": contract.ProcessedSource,
		})
	}
	input := map[string]any{
		"contracts": payloadContracts,
	}
	inputBytes, _ := json.Marshal(input)

	cmd := exec.Command(
		pythonExecutable(),
		filepath.ToSlash(filepath.Join("..", "python_scripts", "run_multi_victim_robustness.py")),
		"--model_dir", modelDir,
		"--target_vuln_type", targetVulnType,
		"--victim_models", strings.Join(victimModels, ","),
		"--variants", fmt.Sprintf("%d", variantsPerSource),
	)
	if strings.TrimSpace(promptText) != "" {
		cmd.Args = append(cmd.Args, "--prompt_text", promptText)
	}
	cmd.Stdin = bytes.NewReader(inputBytes)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("多受害模型鲁棒性评估失败: %w; stderr=%s", err, strings.TrimSpace(stderr.String()))
	}

	var metrics map[string]any
	if err := json.Unmarshal(output, &metrics); err != nil {
		return nil, fmt.Errorf("解析多受害模型评估结果失败: %w; stdout=%s; stderr=%s", err, strings.TrimSpace(string(output)), strings.TrimSpace(stderr.String()))
	}
	return metrics, nil
}

func normalizeRobustStrategy(strategy string) string {
	if strings.TrimSpace(strategy) == "" {
		return RobustStrategyDIPAttack
	}
	return strings.TrimSpace(strategy)
}

func (s *RobustService) loadModelMetadata(artifact string) (*modelArtifactMetadata, error) {
	modelDir, err := resolveArtifactPath(artifact)
	if err != nil {
		return nil, err
	}
	metadataPath := filepath.Join(modelDir, "metadata.json")
	var meta modelArtifactMetadata
	if err := readJSONFile(metadataPath, &meta); err != nil {
		return nil, fmt.Errorf("读取模型元数据失败: %w", err)
	}
	if strings.TrimSpace(meta.TargetVulnType) == "" {
		return nil, fmt.Errorf("模型元数据缺少 target_vuln_type")
	}
	return &meta, nil
}

func scoreForLabel(result *modelInferenceResult, label string) float64 {
	if result == nil {
		return 0
	}
	if result.Scores != nil {
		if v, ok := result.Scores[label]; ok {
			return v
		}
	}
	normalized := strings.ToLower(strings.ReplaceAll(label, "-", "_"))
	for _, item := range result.TopK {
		if strings.ToLower(strings.ReplaceAll(item.Token, "-", "_")) == normalized {
			return item.Score
		}
	}
	return 0
}

func averageConfidenceDrop(rows []robustPerContract) float64 {
	total := 0.0
	count := 0
	for _, row := range rows {
		if row.AdvTotal == 0 {
			continue
		}
		total += row.AvgConfDrop
		count++
	}
	return safeDiv(total, float64(maxInt(count, 1)))
}

func averageQueries(rows []robustPerContract) float64 {
	total := 0.0
	count := 0
	for _, row := range rows {
		if row.AdvTotal == 0 {
			continue
		}
		total += row.AvgQueries
		count++
	}
	return safeDiv(total, float64(maxInt(count, 1)))
}

func averagePerturbationRate(rows []robustPerContract) float64 {
	total := 0.0
	count := 0
	for _, row := range rows {
		if row.AdvTotal == 0 {
			continue
		}
		total += row.AvgPerturbationRate
		count++
	}
	return safeDiv(total, float64(maxInt(count, 1)))
}

func averageVisiblePerturbationRate(rows []robustPerContract) float64 {
	total := 0.0
	count := 0
	for _, row := range rows {
		if row.AdvTotal == 0 {
			continue
		}
		total += row.AvgVisiblePerturbationRate
		count++
	}
	return safeDiv(total, float64(maxInt(count, 1)))
}

func averageCodeBLEU(rows []robustPerContract) float64 {
	total := 0.0
	count := 0
	for _, row := range rows {
		if row.AdvTotal == 0 {
			continue
		}
		total += row.AvgCodeBLEU
		count++
	}
	return safeDiv(total, float64(maxInt(count, 1)))
}

func totalQueryBudgetHits(rows []robustPerContract) int {
	total := 0
	for _, row := range rows {
		total += row.QueryBudgetHits
	}
	return total
}

func buildVisibilityWarning(rows []robustPerContract) string {
	if len(rows) == 0 {
		return ""
	}
	totalVisible := 0.0
	totalPerturb := 0.0
	count := 0
	for _, row := range rows {
		if row.AdvTotal == 0 {
			continue
		}
		totalVisible += row.AvgVisiblePerturbationRate
		totalPerturb += row.AvgPerturbationRate
		count++
	}
	if count == 0 {
		return ""
	}
	avgVisible := safeDiv(totalVisible, float64(count))
	avgPerturb := safeDiv(totalPerturb, float64(count))
	if avgPerturb >= 0.2 && avgVisible <= 0.02 {
		return "本次攻击的大部分改动未进入模型实际可见的输入窗口，当前 0% 结果更偏向“攻击未有效命中模型输入”，不宜直接解释为模型鲁棒性强。"
	}
	return ""
}

func preferSuccessAverage(successSum float64, successCount int, allSum float64, allCount int) float64 {
	if successCount > 0 {
		return safeDiv(successSum, float64(successCount))
	}
	return safeDiv(allSum, float64(maxInt(allCount, 1)))
}

func mustMarshal(v any) string {
	data, _ := json.Marshal(v)
	return string(data)
}

func readJSONFile(path string, out any) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, out)
}

func safeDiv(numerator, denominator float64) float64 {
	if denominator == 0 {
		return 0
	}
	return numerator / denominator
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func (s *RobustService) fail(jobID string, err error) {
	finish := time.Now()
	_ = s.DB.Model(&model.RobustJob{}).Where("id = ?", jobID).Updates(map[string]any{
		"status":      model.RobustJobStatusFailed,
		"error":       err.Error(),
		"finished_at": finish,
	}).Error
}
